src/init/features/functions/index.ts

```diff
@@ -30,6 +26,106 @@ export async function doSetup(setup: any, config: any, options: Options) {
       ensure(projectId, "runtimeconfig.googleapis.com", "unused", true),
     ]);
   }
+  setup.functions = {};
+  if (!config.src.functions) {
+    // if functions have not been initialized yet
+    setup.config.functions = [];
+    return codebaseSetup(setup, config, false);
+  }
+  // if functions have already been initialized, make sure config is validated and normalized
+  setup.config.functions = normalizeAndValidate(setup.config.functions);
+
+  logger.info();
+  logger.info("Detected " + clc.bold("existing codebase(s)."));
+  logger.info();
+
+  const choices = [
+    {
+      name: "Initialize",
+      value: "new",
+    },
+    {
+      name: "Re-initialize",
+      value: "reinit",
+    },
+  ];
+  const initOpt = await promptOnce({
+    type: "list",
+    message: "Would you like to initialize a new codebase, or reinitialize an existing one?",
+    default: "new",
+    choices,
+  });
+  return codebaseSetup(setup, config, initOpt === "reinit");
+}
+
+/**
+ *  User dialogue to set up configuration for functions codebase.
+ */
+async function codebaseSetup(setup: any, config: Config, reinit: boolean): Promise<any> {
+  if (reinit) {
+    let codebase;
+    if (setup.config.functions.length > 1) {
+      const choices = setup.config.functions.map((cfg: any) => ({
+        name: cfg["codebase"],
+        value: cfg["codebase"],
+      }));
+      codebase = await promptOnce({
+        type: "list",
+        message: "Which codebase would you like to re-initialize?",
+        choices,
+      });
+    } else {
+      codebase = setup.config.functions[0].codebase; // only one codebase exists
+    }
+
+    const cbconfig = configForCodebase(setup.config.functions, codebase);
+    setup.functions.source = cbconfig.source;
+    setup.functions.codebase = cbconfig.codebase;
+
+    logger.info("\nRe-initializing " + clc.bold(`codebase ${codebase}...\n`));
+  } else {
+    logger.info("Let's create a new codebase for your functions.");
+    logger.info("A directory corresponding to the codebase will be created in your project");
+    logger.info("with sample code pre-configured.\n");
+
+    logger.info("See https://firebase.google.com/docs/functions/organize-functions for");
+    logger.info("more information on organizing your functions using codebases.\n");
+
+    logger.info("Functions can be deployed with " + clc.bold("firebase deploy") + ".\n");
+
+    const source = await promptOnce({
+      type: "input",
+      message: "Where would you like to initialize your new functions source?",
+      default: "functions",
+    });
+    const codebase = await promptOnce({
```

Comment:
It looks like we never examine the user-input for codebase. We should make sure user input is a valid codebase name, and if now prompt user to try again.

You might consider something like this:

```
    let attempts = 0
    while (true) {
      if (attempts >= MAX_ATTEMPTS) {
        throw new FirebaseError("Exceeded max attempt to input valid codebase name.");
      }
      prefix = await promptForCodebase(errMsg);
      attempts += 1;
    }
```

Reply:
So my thought process was that `normalizeAndValidate(config)` will check the user-input for me when validating the config object (with the user-input codebase name inserted inside it). If `normalizeAndValidate` fails, then the error is logged and `initNewCodebase` is called recursively and the developer must try to input another codebase name. This may be too subtle of a way to check user input; I can see how explicitly checking the user-input codebase name would be more clear.